File: app/src/main/java/com/tool/tree/KrScriptConfig.java
package com.tool.tree;

import android.content.Context;

import com.omarea.krscript.executor.ScriptEnvironmen;
import com.omarea.krscript.model.PageNode;

import java.io.InputStream;
import java.nio.charset.Charset;
import java.util.HashMap;

public class KrScriptConfig {
    private static final String ASSETS_FILE = "file:///android_asset/";

    private static final String TOOLKIT_DIR = "toolkit_dir";
    private static final String TOOLKIT_DIR_DEFAULT = "file:///android_asset/home";

    private static final String EXECUTOR_CORE = "executor_core";
    private static final String PAGE_LIST_CONFIG = "page_list_config";
    private static final String PAGE_LIST_CONFIG_SH = "page_list_config_sh";
    private static final String FAVORITE_CONFIG = "favorite_config";
    private static final String FAVORITE_CONFIG_SH = "favorite_config_sh";
    private static final String CUSTOM_TAB3_CONFIG = "custom_tab3_config";
    private static final String CUSTOM_TAB3_CONFIG_SH = "custom_tab3_config_sh";
    private static final String CUSTOM_TAB4_CONFIG = "custom_tab4_config";
    private static final String CUSTOM_TAB4_CONFIG_SH = "custom_tab4_config_sh";
    private static final String BEFORE_START_SH = "before_start_sh";

    private static HashMap<String, String> configInfo;

    private final String EXECUTOR_CORE_DEFAULT = "file:///android_asset/root/executor.sh";
    private final String PAGE_LIST_CONFIG_DEFAULT = "file:///android_asset/root/more.xml";
    private final String FAVORITE_CONFIG_DEFAULT = "file:///android_asset/root/favorites.xml";
    private final String CUSTOM_TAB3_DEFAULT = "file:///android_asset/root/tab3.xml";
    private final String CUSTOM_TAB4_DEFAULT = "file:///android_asset/root/tab4.xml";
    private final String BEFORE_START_SH_DEFAULT = "";

    public KrScriptConfig init(Context context) {
        if (configInfo == null) {
            configInfo = new HashMap<>();
            configInfo.put(EXECUTOR_CORE, EXECUTOR_CORE_DEFAULT);
            configInfo.put(TOOLKIT_DIR, TOOLKIT_DIR_DEFAULT);
            configInfo.put(BEFORE_START_SH, BEFORE_START_SH_DEFAULT);

            try {
                String fileName = context.getString(R.string.kr_script_config);
                if (fileName.startsWith(ASSETS_FILE)) {
                    fileName = fileName.substring(ASSETS_FILE.length());
                }
                InputStream inputStream = context.getAssets().open(fileName);
                byte[] bytes = new byte[inputStream.available()];
                inputStream.read(bytes);
                inputStream.close();

                String[] rows = new String(bytes, Charset.defaultCharset()).split("\n");
                for (String row : rows) {
                    String rowText = row.trim();
                    if (!rowText.startsWith("#") && rowText.contains("=")) {
                        int separator = rowText.indexOf('=');
                        String key = rowText.substring(0, separator).trim();
                        String value = rowText.substring(separator + 2, rowText.length() - 1).trim();
                        configInfo.put(key, value);
                    }
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }

            ScriptEnvironmen.init(context, getExecutorCore(), getToolkitDir());
        }
        return this;
    }

    public HashMap<String, String> getVariables() {
        return configInfo;
    }

    private String getExecutorCore() {
        return configInfo != null && configInfo.containsKey(EXECUTOR_CORE)
                ? configInfo.get(EXECUTOR_CORE) : EXECUTOR_CORE_DEFAULT;
    }

    private String getToolkitDir() {
        return configInfo != null && configInfo.containsKey(TOOLKIT_DIR)
                ? configInfo.get(TOOLKIT_DIR) : TOOLKIT_DIR_DEFAULT;
    }

    private PageNode getPage(String xmlKey, String shKey) {
        if (configInfo == null) return null;
        boolean hasXml = configInfo.containsKey(xmlKey);
        boolean hasSh = configInfo.containsKey(shKey);
        if (!hasXml && !hasSh) return null;
        PageNode page = new PageNode("");
        if (hasSh) page.setPageConfigSh(configInfo.get(shKey));
        if (hasXml) page.setPageConfigPath(configInfo.get(xmlKey));
        return page;
    }

    public PageNode getPageListConfig() { return getPage(PAGE_LIST_CONFIG, PAGE_LIST_CONFIG_SH); }
    public PageNode getFavoriteConfig() { return getPage(FAVORITE_CONFIG, FAVORITE_CONFIG_SH); }
    public PageNode getCustomTab3Config() { return getPage(CUSTOM_TAB3_CONFIG, CUSTOM_TAB3_CONFIG_SH); }
    public PageNode getCustomTab4Config() { return getPage(CUSTOM_TAB4_CONFIG, CUSTOM_TAB4_CONFIG_SH); }

    public String getBeforeStartSh() {
        return configInfo != null && configInfo.containsKey(BEFORE_START_SH)
                ? configInfo.get(BEFORE_START_SH) : BEFORE_START_SH_DEFAULT;
    }
}

File: app/src/main/kotlin/com/tool/tree/KrScriptConfig.kt
package com.tool.tree

import android.content.Context
import com.omarea.krscript.executor.ScriptEnvironmen
import com.omarea.krscript.model.PageNode
import java.nio.charset.Charset

class KrScriptConfig {
    private val EXECUTOR_CORE_DEFAULT = "file:///android_asset/root/executor.sh"
    private val PAGE_LIST_CONFIG_DEFAULT = "file:///android_asset/root/more.toml"
    private val FAVORITE_CONFIG_DEFAULT = "file:///android_asset/root/favorites.toml"
    private val CUSTOM_TAB3_DEFAULT = "file:///android_asset/root/tab3.toml"
    private val CUSTOM_TAB4_DEFAULT = "file:///android_asset/root/tab4.toml"
    private val BEFORE_START_SH_DEFAULT = ""

    fun init(context: Context): KrScriptConfig {
        if (configInfo == null) {
            val info = HashMap<String, String>()
            info[EXECUTOR_CORE] = EXECUTOR_CORE_DEFAULT
            info[PAGE_LIST_CONFIG] = PAGE_LIST_CONFIG_DEFAULT
            info[FAVORITE_CONFIG] = FAVORITE_CONFIG_DEFAULT
            info[CUSTOM_TAB3_CONFIG] = CUSTOM_TAB3_DEFAULT
            info[CUSTOM_TAB4_CONFIG] = CUSTOM_TAB4_DEFAULT
            info[TOOLKIT_DIR] = TOOLKIT_DIR_DEFAULT
            info[BEFORE_START_SH] = BEFORE_START_SH_DEFAULT
            configInfo = info

            try {
                var fileName = context.getString(R.string.kr_script_config)
                if (fileName.startsWith(ASSETS_FILE)) {
                    fileName = fileName.substring(ASSETS_FILE.length)
                }
                val inputStream = context.assets.open(fileName)
                val bytes = ByteArray(inputStream.available())
                inputStream.read(bytes)
                val rows = String(bytes, Charset.defaultCharset()).split("\n")
                for (row in rows) {
                    val rowText = row.trim()
                    if (!rowText.startsWith("#") && rowText.contains("=")) {
                        val separator = rowText.indexOf("=")
                        val key = rowText.substring(0, separator).trim()
                        val value = rowText.substring(separator + 2, rowText.length - 1).trim()
                        info.remove(key)
                        info[key] = value
                    }
                }
            } catch (ex: Exception) {
                ex.printStackTrace()
            }

            ScriptEnvironmen.init(context, getExecutorCore(), getToolkitDir())
        }

        return this
    }

    fun getVariables(): HashMap<String, String>? {
        return configInfo
    }

    private fun getExecutorCore(): String {
        val info = configInfo
        if (info != null && info.containsKey(EXECUTOR_CORE)) {
            return info[EXECUTOR_CORE] ?: EXECUTOR_CORE_DEFAULT
        }
        return EXECUTOR_CORE_DEFAULT
    }

    private fun getToolkitDir(): String {
        val info = configInfo
        if (info != null && info.containsKey(TOOLKIT_DIR)) {
            return info[TOOLKIT_DIR] ?: TOOLKIT_DIR_DEFAULT
        }
        return TOOLKIT_DIR_DEFAULT
    }

    fun getPageListConfig(): PageNode {
        val info = configInfo
        if (info != null) {
            val pageInfo = PageNode("")
            if (info.containsKey(PAGE_LIST_CONFIG_SH)) {
                pageInfo.pageConfigSh = info[PAGE_LIST_CONFIG_SH] ?: ""
            }
            if (info.containsKey(PAGE_LIST_CONFIG)) {
                pageInfo.pageConfigPath = info[PAGE_LIST_CONFIG] ?: ""
            }
            return pageInfo
        }
        return PageNode("")
    }

    fun getFavoriteConfig(): PageNode {
        val info = configInfo
        if (info != null) {
            val pageInfo = PageNode("")
            if (info.containsKey(FAVORITE_CONFIG_SH)) {
                pageInfo.pageConfigSh = info[FAVORITE_CONFIG_SH] ?: ""
            }
            if (info.containsKey(FAVORITE_CONFIG)) {
                pageInfo.pageConfigPath = info[FAVORITE_CONFIG] ?: ""
            }
            return pageInfo
        }
        return PageNode("")
    }

    fun getCustomTab3Config(): PageNode {
        val info = configInfo
        if (info != null) {
            val pageInfo = PageNode("")
            if (info.containsKey("custom_tab3_config_sh")) {
                pageInfo.pageConfigSh = info["custom_tab3_config_sh"] ?: ""
            }
            if (info.containsKey(CUSTOM_TAB3_CONFIG)) {
                pageInfo.pageConfigPath = info[CUSTOM_TAB3_CONFIG] ?: ""
            }
            return pageInfo
        }
        return PageNode("")
    }

    fun getCustomTab4Config(): PageNode {
        val info = configInfo
        if (info != null) {
            val pageInfo = PageNode("")
            if (info.containsKey("custom_tab4_config_sh")) {
                pageInfo.pageConfigSh = info["custom_tab4_config_sh"] ?: ""
            }
            if (info.containsKey(CUSTOM_TAB4_CONFIG)) {
                pageInfo.pageConfigPath = info[CUSTOM_TAB4_CONFIG] ?: ""
            }
            return pageInfo
        }
        return PageNode("")
    }

    fun getBeforeStartSh(): String {
        val info = configInfo
        if (info != null && info.containsKey(BEFORE_START_SH)) {
            return info[BEFORE_START_SH] ?: BEFORE_START_SH_DEFAULT
        }
        return BEFORE_START_SH_DEFAULT
    }

    companion object {
        private const val ASSETS_FILE = "file:///android_asset/"

        private const val TOOLKIT_DIR = "toolkit_dir"
        private const val TOOLKIT_DIR_DEFAULT = "file:///android_asset/home"

        private const val EXECUTOR_CORE = "executor_core"
        private const val PAGE_LIST_CONFIG = "page_list_config"
        private const val PAGE_LIST_CONFIG_SH = "page_list_config_sh"
        private const val FAVORITE_CONFIG = "favorite_config"
        private const val FAVORITE_CONFIG_SH = "favorite_config_sh"
        private const val CUSTOM_TAB3_CONFIG = "custom_tab3_config"
        private const val CUSTOM_TAB3_CONFIG_SH = "custom_tab3_config_sh"
        private const val CUSTOM_TAB4_CONFIG = "custom_tab4_config"
        private const val CUSTOM_TAB4_CONFIG_SH = "custom_tab4_config_sh"
        private const val BEFORE_START_SH = "before_start_sh"

        private var configInfo: HashMap<String, String>? = null
    }
}
